File: Problem13_IsomorphicStrings.java
// Time Complexity : O(n)
// Space Complexity : O(n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public boolean isIsomorphic(String s, String t) {
        HashMap<Character, Character> map = new HashMap<>();
        for(int i = 0; i < s.length(); i++) {
            if(map.containsKey(s.charAt(i))) {
                if(map.get(s.charAt(i)) != t.charAt(i))
                    return false;
            } else {
                if(map.containsValue(t.charAt(i)))
                    return false;
                map.put(s.charAt(i), t.charAt(i));
            }
        }
        return true;
    }
}

File: Problem 13 Isomorphic Strings.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n)
# Space Complexity: O(1)
# Logic: create 2 dict & create key-value of s in 1 & reverse i.e. value key in other dict. traverse through the
# elements if keys are not present in both, store them else check if values are correct or not.

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        dict_s_t = {}
        dict_t_s = {}
        if len(s) != len(t):
            return False
        for i in range(len(s)):
            if s[i] not in dict_s_t and t[i] not in dict_t_s:
                dict_s_t[s[i]] = t[i]
                dict_t_s[t[i]] = s[i]
            else:
                if (s[i] in dict_s_t and dict_s_t[s[i]] != t[i]) or (t[i] in dict_t_s and dict_t_s[t[i]] != s[i]):
                    return False
        return True
